src/julia.h

```diff
@@ -397,6 +397,7 @@ typedef struct {
     unsigned exportp:1;
     unsigned imported:1;
     unsigned deprecated:1;
+    jl_sym_t *deprecation_preferred_name;  // non-NULL when new type incompatible with old
```

Comment:
I think this can just replace `b->deprecated`. IIUC, the original concern is memory consumption of a significantly larger `jl_binding_t*`.